ext/Pod-Html/t/pod2html-lib.pl

```diff
@@ -41,20 +45,40 @@ sub convert_n_test {
     my $infile   = catpath $vol, $new_dir, "$podfile.pod";
     my $outfile  = catpath $vol, $new_dir, "$podfile.html";
 
-    my %default_args_table = (
+    my %args_table = (
         infile      =>    $infile,
         outfile     =>    $outfile,
         podpath     =>    't',
         htmlroot    =>    '/',
         podroot     =>    $cwd,
     );
-    my @default_args_list = map { "--${_}=" . $default_args_table{$_} }
-        keys %default_args_table;
-    # To add/modify args to p2h, use @p2h_args
-    Pod::Html::pod2html(
-        @default_args_list,
-        @p2h_args,
-    );
+    my %no_arg_switches = map { substr($_,2) => 1 } ( qw|
+           --flush --recurse --norecurse
+           --quiet --noquiet --verbose --noverbose
+           --index --noindex --backlink --nobacklink
+           --header --noheader --poderrors --nopoderrors
+    | );
+    if (defined $p2h_args_ref) {
+        for my $sw (keys %{$p2h_args_ref}) {
+            if ($no_arg_switches{$sw}) {
+                $args_table{$sw} = undef;
+            }
+            else {
+                $args_table{$sw} = $p2h_args_ref->{$sw};
+            }
+        }
+    }
+    my @args_list = ();
+    for my $k (keys %args_table) {
+        if (defined $args_table{$k}) {
+            push @args_list, "--" . $k . "=" . $args_table{$k};
+        }
+        else {
+            push @args_list, "--" . $k;
+        }
+    }
+
```

Comment:
Here is perhaps my only _substantive_ comment:

You have `%no_arg_switches`, which is used to replace any given value for an option in the p2h_args_ref.  Then you build up `@args_list` by treating an undef value as `--foo` rather than `--foo=bar`.  This seems like it is likely to be correct in all existing cases (but I didn't verify), *but* it seems a bit backward to me to do this as opposed to checking it very late.  p2h_args_ref would use TRUE to mean "include this no arg switch" and FALSE to exclude it.  This would make it possible for the defaults, later, to have a `--foo` switch that could be removed.  Then, in the loop at line 72, instead of including an argument based on the definedness of the value in args_table, you'd include an argument based on whether the key is in the `no_arg_switches` lookup.

This isn't a big deal.  I think the code is correct, now.  I think it could later impede further refactoring or updating of this code, just a little, but also that is not a very likely event to plan for.

So, if you think it's worth making the change I suggest:  now you have heard the suggestion and can do so.  If you don't, I think the code as presented is likely fine.  Also, we can refactor more later, if needed.

Reply:
> You have `%no_arg_switches`, which is used to replace any given value for an option in the p2h_args_ref. Then you build up `@args_list` by treating an undef value as `--foo` rather than `--foo=bar`. This seems like it is likely to be correct in all existing cases (but I didn't verify), _but_ it seems a bit backward to me to do this as opposed to checking it very late. p2h_args_ref would use TRUE to mean "include this no arg switch" and FALSE to exclude it. This would make it possible for the defaults, later, to have a `--foo` switch that could be removed. Then, in the loop at line 72, instead of including an argument based on the definedness of the value in args_table, you'd include an argument based on whether the key is in the `no_arg_switches` lookup.
> [snip]

My objectives in this refactoring were and are:

* To make `convert_n_test()`'s interface somewhat more sane, by making its third argument a hashref and by keying that hashref on ordinary strings rather than strings that look like command-line switches.

* To make the list of arguments we pass to `Pod::Html::pod2html()` very debuggable.

I believe the first objective has been met.

For the second objective what I want is to be able to debug any call to `convert_n_test()` like this:
```
# %args_table previously populated
print STDERR Dumper(\%args_table);

    my @args_list = ();
    for my $k (keys %args_table) {
        if (defined $args_table{$k}) {
            push @args_list, "--" . $k . "=" . $args_table{$k};
        } else {
            push @args_list, "--" . $k;
        }
    }

print STDERR Dumper(\@args_list);

    Pod::Html::pod2html( @args_list );
```
That is, I want to be able to dump the data being passed to `pod2html()` in either of two ways:  in the form of a lookup table, and/or in the form of the actual list of 'switch-like' strings being bassed to `pod2html()`.  From your verbal description alone I don't see how your suggestion satisfies those objectives.  Could you provide some code?

Thank you very much.
Jim Keenan

Reply:
```
warn Dumper(\%args) if $DEBUGGING;

my @arg_list;
for my $k (keys %arg) {
  if ($is_bool_option{$k}) {
    push @arg_list, "--$k";
  } else {
    push @arg_list, "--$k=$arg{$k}";
  }
}

warn Dumper(\@arg_list) if $DEBUGGING;

Pod::Html::pod2html(@arg_list);
```

Reply:
> ```
> warn Dumper(\%args) if $DEBUGGING;
> 
> my @arg_list;
> for my $k (keys %arg) {
>   if ($is_bool_option{$k}) {
>     push @arg_list, "--$k";
>   } else {
>     push @arg_list, "--$k=$arg{$k}";
>   }
> }
> 
> warn Dumper(\@arg_list) if $DEBUGGING;
> 
> Pod::Html::pod2html(@arg_list);
> ```

@rjbs, to the extent that I understand your proposed alternative -- and I grant that I may not -- I disagree with it.

First, you've renamed some of the variables, which tends to obscure the real changes you're suggesting.  Can I assume that your `%is_bool_option` is the same as my `no_arg_switches`?

Can I assume that your `@arg_list` is the same as my `@args_list`?

And now, most crucially, you are using `%args`.  Is it the same as my `%args_table`?  I don't think so.

The point in `%args_table` is to clearly differentiate the arguments being passed to `convert_n_test()` from those ultimately passed to `Pod::Html::pod2html()` inside `convert_n_test()`.  You haven't show how you would populate `%args`, but I suspect it would blur that distinction.

Currently (*i.e.,* in blead), `convert_n_test()` is structured like this:
```
sub convert_n_test {
    my($podfile, $testname, @p2h_args) = @_;
    # where @p2h_args is a list of strings which are written as if they were command-line options
    ...
    
    # define default values
    ...
    
        Pod::Html::pod2html(
        "--infile=$infile",
        "--outfile=$outfile",
        "--podpath=t",
        "--htmlroot=/",
        "--podroot=$cwd",
        @p2h_args,
    );
```
`Pod::Html::pod2html()` takes a list of strings which are written as if they were command-line options.  I think that's a bad interface design.  It should have been a hash or hashref -- and command-line utility `pod2html` should have used `Getopt::Long` internally to create that hash.  But that horse left the barn in the last century.

What really aggravates me is that `convert_n_test()` in blead mushes together (a) default values which it sets internally and (b) arguments passed to `convert_n_test()`.  My refactoring clearly distinguishes the two.  Since I don't know how you propose to populate `%args`, I don't understand how you can maintain that distinction.

Reply:
Yes, I think %arg was the same as %arg_table.  I just tried to bang out sufficiently demonstrative code to show what I meant about passing a value to a `--switch` based on something other than the definedness of the hash entry, and to get it done before falling asleep last night.

You're storing "undef" to mean that there's no value.  My suggestion was that instead, I would stick to using a separate list of "this switch is only present or not present".  That way you'd never have a bug where the "infile" option ended up undef, and so `--infile` was passed as a switch with no argument.  Or you could notice an error where some no-arg switch (like `--quiet`) had gotten its value set to 2.  There are some benefits of this which I alluded to.

If you don't agree, it's fine, I think your code is fine to apply as-is (as I said initially).